### PR TITLE
microbin: add service

### PR DESCRIPTION
**Pull request by surfaceflinger**
## Description of changes

Added a NixOS module for microbin.
pinging @dit7ya and @figsoda as they both maintain microbin package :)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by tomfitzhenry — CHANGES_REQUESTED** [review]
+1 to the package and the systemd hardening.

Is there a way to configure this module in a [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) style, rather than introducing options for each configuration variable?

**Comment by surfaceflinger** [comment]
> +1 to the package and the systemd hardening.
> 
> Is there a way to configure this module in a [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) style, rather than introducing so many options?

thanks i'll check this out after work today 🙂
personally I just wanted to have everything exposed as options because most people just take a look at search.nixos.org and it would be nicely listed there

**Comment by surfaceflinger** [comment]
@tomfitzhenry can you check if it's fine now? It could be more minimal by getting rid of the envFile like here https://github.com/NixOS/nixpkgs/blob/85d77bc124df88efe65d0e8879ac55be1bd61d7f/nixos/modules/services/monitoring/uptime-kuma.nix
but then things like MICROBIN_PORT would have to be explicitly set as strings which kinda stinks imo 😕

**Comment by TLATER** [comment]
I don't think there's a particular reason to get rid of the env file. I didn't know `EnvironmentFile` could be repeated, but since it can be, this isn't in the way of user overrides.

If you *wanted* to, I think you could eat your cake and have it too by doing something like this:

```nix
Environment = lib.mapAttrs (_: val: 
  if lib.isBool val
  then lib.boolToString val
  else toString val;
) cfg.settings;
```

I have two nits:
- Some of the systemd hardening options you set are implied by `DynamicUser`, and therefore superfluous, but this isn't harmful.
- Maybe add a note to the settings description pointing users to the secrets option, to make sure nobody puts secrets in `settings`: https://github.com/NixOS/nixpkgs/pull/251996/files#diff-c68102ca921100d399d2be2f558f21d33f74a81559d2366003087365f4fe0e80R36

Otherwise lgtm!

**Comment by surfaceflinger** [comment]
@TLATER can you take a look again?

**Comment by TLATER** [comment]
Nice! Didn't expect you to address the nits. I also like the detail of sorting the env variables, and good spot on the `nullOr`. I have nothing else to add.

**Review by tomfitzhenry — APPROVED** [review]
LGTM, other than the minor issues identified. Nice work on reworking to RFC42-style! :)

File: nixos/modules/services/web-apps/microbin.nix [hunk 1]
```diff
@@ -0,0 +1,99 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.microbin;
+in
+{
+  meta.maintainers = with lib.maintainers; [ surfaceflinger ];
```
[the remark this answers is not in the document]
Comment (surfaceflinger) [reply nixos/modules/services/web-apps/microbin.nix]: Corrected

File: nixos/modules/services/web-apps/microbin.nix [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  config = lib.mkIf cfg.enable {
+    services.microbin.settings = with lib; {
+      MICROBIN_BIND = mkDefault "0.0.0.0";
+      MICROBIN_DISABLE_TELEMETRY = mkDefault true;
+      MICROBIN_NO_LISTING = mkDefault true;
+      MICROBIN_PORT = mkDefault "8080";
+    };
+
+    systemd.services.microbin = {
+      after = [ "network-online.target" ];
```
[the remark this answers is not in the document]
Comment (surfaceflinger) [reply nixos/modules/services/web-apps/microbin.nix]: Done, thanks

File: nixos/modules/services/web-apps/microbin.nix [hunk 3]
```diff
@@ -0,0 +1,99 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.microbin;
+in
+{
+  options.services.microbin = {
+    enable = lib.mkEnableOption (lib.mdDoc "MicroBin is a super tiny, feature rich, configurable, self-contained and self-hosted paste bin web application.");
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/microbin.nix:8]:
```suggestion
    enable = lib.mkEnableOption (lib.mdDoc "MicroBin is a super tiny, feature rich, configurable paste bin web application");
```

File: nixos/modules/services/web-apps/microbin.nix [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    package = lib.mkOption {
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/microbin.nix:10]: Please use mkPackageOption

File: nixos/modules/services/web-apps/microbin.nix [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      description = lib.mdDoc "Default data folder for MicroBin.";
+    };
+
+    passwordFile = lib.mkOption {
+      type = lib.types.nullOr lib.types.path;
+      default = null;
+      example = "/run/secrets/microbin.env";
+      description = lib.mdDoc ''
+        Additional environment variables containing credentials.
+        Possible secret variables are
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/microbin.nix:45]: This should describe the format and is not really limited to those values. I could set any other, right?

File: nixos/modules/services/web-apps/microbin.nix [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      MICROBIN_DISABLE_TELEMETRY = mkDefault true;
+      MICROBIN_LIST_SERVER = mkDefault false;
+      MICROBIN_PORT = mkDefault "8080";
+    };
+
+    systemd.services.microbin = {
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      startLimitIntervalSec = 14400;
+      startLimitBurst = 10;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/microbin.nix:67]: Why do we need to increase this?

File: nixos/modules/services/web-apps/microbin.nix [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        PrivateDevices = true;
+        PrivateUsers = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ReadWritePaths = "${cfg.dataDir}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/microbin.nix:86]:
```suggestion
        ReadWritePaths = cfg.dataDir;
```

File: nixos/modules/services/web-apps/microbin.nix [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        StateDirectory = "microbin";
+        SystemCallArchitectures = [ "native" ];
+        SystemCallFilter = [ "@system-service" ];
+        WorkingDirectory = "${cfg.dataDir}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/microbin.nix:93]:
```suggestion
        WorkingDirectory = cfg.dataDir;
```

File: nixos/modules/services/web-apps/microbin.nix (hunk 5 above)
Comment (surfaceflinger) [reply nixos/modules/services/web-apps/microbin.nix]: Please check if the new wording is better

File: nixos/modules/services/web-apps/microbin.nix [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    };
+
+    passwordFile = lib.mkOption {
+      type = lib.types.nullOr lib.types.path;
+      default = null;
+      example = "/run/secrets/microbin.env";
+      description = lib.mdDoc ''
+        Path to file containing environment variables.
+        Useful for passing down secrets.
+        Variables that can be considered secret are:
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/microbin.nix:41]:
```suggestion
        Variables that can be considered secrets are:
```

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

File: nixos/modules/services/web-apps/microbin.nix (hunk 9 above)
Comment (surfaceflinger) [reply nixos/modules/services/web-apps/microbin.nix]: done

**Review by thiagokokada — APPROVED** [review]

**Pull request merged**